x-pack/plugins/cloud_security_posture/public/components/setup_status.tsx

```diff
@@ -0,0 +1,187 @@
+      defaultMessage: 'Cloud Security Posture',
+    })}
+    // TODO: Add real docs link once we have it
+    docsLink={'https://www.elastic.co/guide/index.html'}
+    logo={'logoSecurity'}
+    actions={{
+      elasticAgent: {
+        href: cisIntegrationLink,
+        isDisabled: !cisIntegrationLink,
+        title: i18n.translate('xpack.csp.setupStatus.packageNotInstalled.buttonLabel', {
+          defaultMessage: 'Add a CIS integration',
+        }),
+        description: i18n.translate('xpack.csp.setupStatus.packageNotInstalled.description', {
+          defaultMessage:
+            'Use our CIS Kubernetes Benchmark integration to measure your Kubernetes cluster setup against the CIS recommendations.',
+        }),
+      },
+    }}
+  />
+);
+
+const defaultLoadingRenderer = () => (
+  <CspLoadingState data-test-subj={LOADING_STATE_TEST_SUBJECT}>
+    <FormattedMessage id="xpack.csp.setupStatus.loadingDescription" defaultMessage="Loading..." />
+  </CspLoadingState>
+);
+
+const defaultErrorRenderer = (error: unknown) => (
+  <EuiEmptyPrompt
+    css={css`
+      margin-top: 50px;
+    `}
+    color="danger"
+    iconType="alert"
+    data-test-subj={ERROR_STATE_TEST_SUBJECT}
+    title={
+      <h2>
+        <FormattedMessage
+          id="xpack.csp.setupStatus.errorRenderer.errorTitle"
+          defaultMessage="We couldn't fetch your cloud security posture data"
+        />
+      </h2>
+    }
+    body={
+      isCommonError(error) ? (
+        <p>
+          <FormattedMessage
+            id="xpack.csp.setupStatus.errorRenderer.errorDescription"
+            defaultMessage="{error} {statusCode}: {body}"
+            values={{
+              error: error.body.error,
+              statusCode: error.body.statusCode,
+              body: error.body.message,
+            }}
+          />
+        </p>
+      ) : undefined
+    }
+  />
+);
+
+const defaultNoDataRenderer = () => {
+  return (
+    <NoDataPage
+      data-test-subj={DEFAULT_NO_DATA_TEST_SUBJECT}
+      css={css`
+        margin-top: 50px;
+      `}
+      pageTitle={i18n.translate('xpack.csp.setupStatus.defaultNoDataConfig.pageTitle', {
+        defaultMessage: 'No data found',
+      })}
+      solution={i18n.translate('xpack.csp.setupStatus.defaultNoDataConfig.solutionNameLabel', {
+        defaultMessage: 'Cloud Security Posture',
+      })}
+      // TODO: Add real docs link once we have it
+      docsLink={'https://www.elastic.co/guide/index.html'}
+      logo={'logoSecurity'}
+      actions={{}}
+    />
+  );
+};
+
+interface SetupStatusProps<TData, TError> {
+  children: React.ReactNode;
+  query?: UseQueryResult<TData, TError>;
+  loadingRender?: () => React.ReactNode;
+  errorRender?: (error: TError) => React.ReactNode;
+  noDataRenderer?: () => React.ReactNode;
+}
+
+export const SetupStatus = <TData, TError>({
+  children,
+  query,
+  loadingRender = defaultLoadingRenderer,
+  errorRender = defaultErrorRenderer,
+  noDataRenderer = defaultNoDataRenderer,
```

Comment:
I'm against. removing those means copy-pasting those components into 3 routes plus the function to decide which state to render.
`loading` `error` and `noData` are common to every query, no need to re-write the same logic on every route:

```
qurey.isLoading && <Loader />
qurey.isError && <Error />
!qurey.data && <NoData />

```

if a route demands special treatment there is an option to pass custom renderers or not pass `query` at all and pass your own logic into children.

Reply:
but it would be nice if we didn't have to :) 

what i mean is that maybe the use case for custom rendering of loaders/error/nodata is not needed. for example, the error renderer from the rules page is not currently needed. so maybe we can introduce these APIs when they are needed. unless they are 😄